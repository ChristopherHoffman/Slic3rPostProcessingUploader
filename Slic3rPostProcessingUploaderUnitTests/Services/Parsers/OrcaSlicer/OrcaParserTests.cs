using Slic3rPostProcessingUploader.Services.Parsers.AnycubicSlicerNext;
using Slic3rPostProcessingUploader.Services.Parsers.OrcaSlicer;
using Slic3rPostProcessingUploaderUnitTests.Services.Parsers.AnycubicSlicerNext;
using Snapshooter.MSTest;


namespace Slic3rPostProcessingUploaderUnitTests.Services.Parsers.OrcaSlicer
{
    [TestClass]
    public class OrcaParserTests
    {
        [TestMethod]
        public void ShouldReturnAnEmptySettingWhenGcodeIsEmpty()
        {
            var parser = new OrcaParser("");
            var result = parser.ParseGcode("");
            Assert.IsNotNull(result);

            Snapshot.Match(result, matchOptions => matchOptions.HashField("settings.Snapshot"));
        }

        [TestMethod]
        public void ShouldReturnExpectedValuesWhenGivenFullGcode()
        {
            var parser = new OrcaParser("");
            var result = parser.ParseGcode(OrcaParserTestGcode.CalibrationCube);

            Snapshot.Match(result, matchOptions => matchOptions.HashField("settings.Snapshot"));
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithNoReplacements()
        {
            string template = "Settings:";

            var parser = new OrcaParser(template);
            var result = parser.ParseGcode(OrcaParserTestGcode.CalibrationCube);

            Assert.AreEqual("Settings:", result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithASingleReplacement()
        {
            string template = """
                Settings:
                    Layer Height: {{layer_height}}
                """;

            var parser = new OrcaParser(template);
            var result = parser.ParseGcode(OrcaParserTestGcode.CalibrationCube);

            Assert.AreEqual("""
                Settings:
                    Layer Height: 0.2
                """, result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithMultipleReplacements()
        {
            string template = """
                Settings:
                    Layer Height: {{layer_height}}
                    First Layer Height: {{first_layer_height}}
                    Wall Loops: {{wall_loops}}
                    Top Shell Layers: {{top_shell_layers}}
                    Bottom Shell Layers: {{bottom_shell_layers}}
                    Sparse Infill Density: {{sparse_infill_density}}
                """;

            var parser = new OrcaParser(template);
            var result = parser.ParseGcode(OrcaParserTestGcode.CalibrationCube);

            Assert.AreEqual("""
                Settings:
                    Layer Height: 0.2
                    First Layer Height: 0.200
                    Wall Loops: 3
                    Top Shell Layers: 4
                    Bottom Shell Layers: 3
                    Sparse Infill Density: 15%
                """, result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderFullTemplateWhenGivenAGcodeWithTwoFilaments()
        {
            var parser = new OrcaParser("");
            var result = parser.ParseGcode(OrcaParserTestGcode.CalibrationCubeTwoFilament);

            Snapshot.Match(result, matchOptions => matchOptions.HashField("settings.Snapshot"));
        }

    }
}
