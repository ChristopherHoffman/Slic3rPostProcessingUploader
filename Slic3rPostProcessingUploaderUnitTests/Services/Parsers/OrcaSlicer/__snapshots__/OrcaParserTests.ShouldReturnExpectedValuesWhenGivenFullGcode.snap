{
  "Slicer": "OrcaSlicer",
  "CuraVersion": "2.2.0-rc",
  "PluginVersion": null,
  "settings": {
    "note": "Settings:\n\nQuality:\n  Layer Height:\n    Layer Height: 0.2\n    First Layer Height: 0.200\n  Line Width:\n    Default Line Width: 0.4\n  Seam:\n    Seam Position: aligned\n  Precision:\n    Precise Wall: 0\n  Ironing:\n    Ironing: no ironing\n  Wall Generator:\n    Generator: arachne\n  Walls And Surfaces:\n    Only One Wall on Top Surfaces: 0\n\nStrength:\n  Walls:\n    Wall Loops: 3\n  Top/Bottom Shells:\n    Top Shell Layers: 4\n    Top Shell Thickness: 0.8\n    Bottom Shell Layers: 3\n    Bottom Shell Thickness: 0\n  Infill:\n    Sparse Infill Density: 15%\n    Sparse Infill Pattern: crosshatch\n\nSpeed:\n  First Layer Speed:\n    First Layer: 45\n    Number of Slow Layers: 0\n  Other Layers Speed:\n    Outer Wall: 60\n    Inner Wall: 80\n    Sparse Infill: 80\n    Internal Solid Infill: 80\n    Top Surface: 50\n  Travel Speed:\n    Travel: 150\n\nSupport:\n  Support:\n    Enable Support: 0\n    Type: normal(auto)\n    Style: default\n    Threshold Angle: 30\n    On Build Plate Only: 0\n  Raft:\n    Raft Layers: 0\n\nMultimaterial:\n  Prime Tower:\n    Enable: 0\n    Width: 60\n    Brim Width: 3\n  Ooze Prevention:\n    Enable: 0\n  Flush Options:\n    Flush into Objects' Infill: 0\n    Flush into Objects' Support: 1\n  Advanced:\n    Use Beam Interlocking: 0\n\nOther:\n  Skirt:\n    Skirt Type: combined\n    Skirt Loops: 0\n    Skirt Min Extrusion Length: 4\n  Brim:\n    Brim Type: auto_brim\n    Brim Width: 5\n  Special Mode:\n    Slicing Mode: regular\n    Print Sequence: by layer\n    Spiral Vase: 0\n    Fuzzy Skin: none",
    "print_name": null,
    "time_stamp": null,
    "estimated_print_time_seconds": 1319,
    "material_used_mg": null,
    "Snapshot": "R9jFouPRGj/K4wgBU70/9vEAcaqZnmN0y9L6dsrTCQw=",
    "file_name": null,
    "filamentUsage": [
      {
        "Id": null,
        "Filament": {
          "Id": "00000000-0000-0000-0000-000000000000",
          "DisplayName": "Other"
        },
        "AmountMg": null,
        "LengthInM": null,
        "VolumeMl": null,
        "EstimatedAmountMg": null,
        "EstimatedLengthInM": 1.52,
        "EstimatedVolumeMl": null,
        "Source": "Length",
        "EstimatedSource": "Length",
        "Notes": ""
      }
    ]
  }
}
